File: Proyecto/Calendario_App/src/main/java/org/example/calendario_app/dao/impl/EventoDAOImpl.java
package org.example.calendario_app.dao.impl;

import org.example.calendario_app.dao.EventoDAO;
import org.example.calendario_app.model.Evento;
import org.example.calendario_app.util.DatabaseConnection;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class EventoDAOImpl implements EventoDAO {
    private final DatabaseConnection databaseConnection;

    public EventoDAOImpl() {
        this.databaseConnection = new DatabaseConnection();
    }

    @Override
    public List<Evento> findAllByUsuarioId(int idUsuario) {
        List<Evento> eventos = new ArrayList<>();
        // Query to get:
        // 1. Events created by the user
        // 2. Events that have a label which is shared with the user (via
        // etiquetas_usuarios)
        // 3. Events that have a label which is shared with a group the user is in (via
        // etiquetas_grupos -> grupos_usuarios)

        // Query to get:
        // 1. Events created by the user
        // 2. Events that have a label which is shared with the user (via
        // etiquetas_usuarios)

        String query = """
                    SELECT DISTINCT ev.*
                    FROM eventos ev
                    LEFT JOIN etiquetas e ON ev.id_etiqueta = e.id_etiqueta
                    LEFT JOIN etiquetas_usuarios eu ON e.id_etiqueta = eu.id_etiqueta
                    LEFT JOIN eventos_usuarios evu ON ev.id_evento = evu.id_evento
                    WHERE ev.id_creador = ?
                       OR eu.id_usuario = ?
                       OR evu.id_usuario = ?
                """;

        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query)) {

            pstmt.setInt(1, idUsuario);
            pstmt.setInt(2, idUsuario);
            pstmt.setInt(3, idUsuario);

            try (ResultSet rs = pstmt.executeQuery()) {
                while (rs.next()) {
                    Timestamp startTs = rs.getTimestamp("fecha_inicio");
                    Timestamp endTs = rs.getTimestamp("fecha_fin");

                    eventos.add(new Evento(
                            rs.getInt("id_evento"),
                            rs.getString("titulo"),
                            rs.getString("descripcion"),
                            startTs != null ? startTs.toLocalDateTime() : null,
                            endTs != null ? endTs.toLocalDateTime() : null,
                            rs.getString("ubicacion"),
                            rs.getInt("id_creador"),
                            (Integer) rs.getObject("id_etiqueta")));
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return eventos;
    }

    @Override
    public int save(Evento evento) {
        String query = "INSERT INTO eventos (titulo, descripcion, fecha_inicio, fecha_fin, ubicacion, id_creador, id_etiqueta) VALUES (?, ?, ?, ?, ?, ?, ?)";
        int idGenerado = -1;

        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {

            System.out.println("Saving Event: Title=" + evento.getTitulo() + ", LabelID=" + evento.getId_etiqueta());

            pstmt.setString(1, evento.getTitulo());
            pstmt.setString(2, evento.getDescripcion());
            pstmt.setTimestamp(3, Timestamp.valueOf(evento.getFecha_inicio()));
            pstmt.setTimestamp(4, evento.getFecha_fin() != null ? Timestamp.valueOf(evento.getFecha_fin()) : null);
            pstmt.setString(5, evento.getUbicacion());
            pstmt.setInt(6, evento.getId_creador());
            if (evento.getId_etiqueta() != null) {
                pstmt.setInt(7, evento.getId_etiqueta());
            } else {
                pstmt.setNull(7, java.sql.Types.INTEGER);
            }

            int affectedRows = pstmt.executeUpdate();

            if (affectedRows > 0) {
                try (ResultSet generatedKeys = pstmt.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        idGenerado = generatedKeys.getInt(1);
                        evento.setId(idGenerado);

                        // Auto-sharing logic (New)
                        if (evento.getId_etiqueta() != null) {
                            shareWithGroupMembers(conn, idGenerado, evento.getId_etiqueta(), evento.getId_creador());
                        }
                    }
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return idGenerado;
    }

    private void shareWithGroupMembers(Connection conn, int eventId, int labelId, int creatorId) throws SQLException {
        // 1. Check if label is linked to a group
        String checkGroupQuery = "SELECT id_grupo FROM etiquetas WHERE id_etiqueta = ?";
        Integer groupId = null;
        try (PreparedStatement pstmt = conn.prepareStatement(checkGroupQuery)) {
            pstmt.setInt(1, labelId);
            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    groupId = (Integer) rs.getObject("id_grupo");
                }
            }
        }

        // 2. If group exists, invite all members (except creator)
        if (groupId != null) {
            String membersQuery = "SELECT id_usuario FROM grupos_usuarios WHERE id_grupo = ? AND id_usuario != ?";
            String inviteQuery = "INSERT INTO eventos_usuarios (id_evento, id_usuario, estado, notificado) VALUES (?, ?, 'pendiente', 0) ON DUPLICATE KEY UPDATE id_usuario=id_usuario";

            try (PreparedStatement pstmtMembers = conn.prepareStatement(membersQuery);
                    PreparedStatement pstmtInvite = conn.prepareStatement(inviteQuery)) {

                pstmtMembers.setInt(1, groupId);
                pstmtMembers.setInt(2, creatorId);

                try (ResultSet rs = pstmtMembers.executeQuery()) {
                    while (rs.next()) {
                        int userId = rs.getInt("id_usuario");
                        pstmtInvite.setInt(1, eventId);
                        pstmtInvite.setInt(2, userId);
                        pstmtInvite.executeUpdate();
                    }
                }
            }
        }
    }

    @Override
    public void delete(int id) {
        String query = "DELETE FROM eventos WHERE id_evento = ?";
        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query)) {
            pstmt.setInt(1, id);
            pstmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: Proyecto_Android/Calendario_Android_App/app/src/main/java/com/example/calendario_android_app/dao/impl/EventoDAOImpl.kt
package com.example.calendario_android_app.dao.impl

import com.example.calendario_android_app.dao.EventoDAO
import com.example.calendario_android_app.model.Evento
import com.example.calendario_android_app.model.EstadoEvento
import com.example.calendario_android_app.util.DatabaseConnection
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.sql.ResultSet
import java.time.LocalDate
import java.time.LocalDateTime

class EventoDAOImpl : EventoDAO {

    override suspend fun getEventosByUsuarioAndFecha(idUsuario: Int, fecha: LocalDate): List<Evento> = withContext(Dispatchers.IO) {
        val eventos = mutableListOf<Evento>()
        val connection = DatabaseConnection.getConnection()
        
        try {
            connection?.let { conn ->
                val sql = """
                    SELECT e.*, et.color 
                    FROM eventos e
                    INNER JOIN eventos_usuarios eu ON e.id_evento = eu.id_evento
                    LEFT JOIN etiquetas et ON e.id_etiqueta = et.id_etiqueta
                    WHERE eu.id_usuario = ? 
                    AND DATE(e.fecha_inicio) = ?
                    ORDER BY e.fecha_inicio ASC
                """
                
                val statement = conn.prepareStatement(sql)
                statement.setInt(1, idUsuario)
                statement.setString(2, fecha.toString())
                
                val resultSet = statement.executeQuery()
                
                while (resultSet.next()) {
                    eventos.add(mapResultSetToEvento(resultSet))
                }
                
                resultSet.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        
        eventos
    }

    override suspend fun getEventosByUsuario(idUsuario: Int): List<Evento> = withContext(Dispatchers.IO) {
        val eventos = mutableListOf<Evento>()
        val connection = DatabaseConnection.getConnection()
        
        try {
            connection?.let { conn ->
                val sql = """
                    SELECT e.*, et.color 
                    FROM eventos e
                    INNER JOIN eventos_usuarios eu ON e.id_evento = eu.id_evento
                    LEFT JOIN etiquetas et ON e.id_etiqueta = et.id_etiqueta
                    WHERE eu.id_usuario = ?
                    ORDER BY e.fecha_inicio ASC
                """
                
                val statement = conn.prepareStatement(sql)
                statement.setInt(1, idUsuario)
                
                val resultSet = statement.executeQuery()
                
                while (resultSet.next()) {
                    eventos.add(mapResultSetToEvento(resultSet))
                }
                
                resultSet.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        
        eventos
    }

    override suspend fun insertEvento(evento: Evento, idUsuario: Int): Boolean {
        return insertEventoReturnId(evento, idUsuario) != -1
    }

    override suspend fun insertEventoReturnId(evento: Evento, idUsuario: Int): Int = withContext(Dispatchers.IO) {
        val connection = DatabaseConnection.getConnection()
        var generatedId = -1
        
        try {
            connection?.let { conn ->
                conn.autoCommit = false // Transaction start
                
                // Ensure tables
                 val createLinkTable = """
                    CREATE TABLE IF NOT EXISTS eventos_usuarios (
                        id_evento INT NOT NULL,
                        id_usuario INT NOT NULL,
                        PRIMARY KEY (id_evento, id_usuario),
                        FOREIGN KEY (id_evento) REFERENCES eventos(id_evento) ON DELETE CASCADE,
                        FOREIGN KEY (id_usuario) REFERENCES usuarios(id_usuario) ON DELETE CASCADE
                    )
                """
                val createStmt = conn.createStatement()
                createStmt.execute(createLinkTable)
                // Add id_etiqueta column if not exists (Primitive migration)
                try {
                     createStmt.execute("ALTER TABLE eventos ADD COLUMN id_etiqueta INT NULL")
                     createStmt.execute("ALTER TABLE eventos ADD CONSTRAINT fk_evento_etiqueta FOREIGN KEY (id_etiqueta) REFERENCES etiquetas(id_etiqueta) ON DELETE SET NULL")
                } catch(e: Exception) {
                    // Ignore column already exists
                }
                createStmt.close()

                val sql = """
                    INSERT INTO eventos (id_creador, id_etiqueta, titulo, descripcion, fecha_inicio, fecha_fin, ubicacion, estado)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                """
                
                val statement = conn.prepareStatement(sql, java.sql.Statement.RETURN_GENERATED_KEYS)
                statement.setObject(1, evento.idCreador)
                statement.setObject(2, evento.idEtiqueta)
                statement.setString(3, evento.titulo)
                statement.setString(4, evento.descripcion)
                statement.setObject(5, evento.fechaInicio)
                statement.setObject(6, evento.fechaFin)
                statement.setString(7, evento.ubicacion)
                statement.setString(8, evento.estado.valor)
                
                val affectedRows = statement.executeUpdate()
                
                if (affectedRows > 0) {
                    val generatedKeys = statement.generatedKeys
                    if (generatedKeys.next()) {
                        generatedId = generatedKeys.getInt(1)
                        
                        // Check if event has a label associated with a group
                        var grupoUsuarios = mutableListOf<Int>()
                        
                        if (evento.idEtiqueta != null) {
                            // Get the group ID for this label
                            val sqlGrupo = "SELECT id_grupo FROM etiquetas WHERE id_etiqueta = ?"
                            val stmtGrupo = conn.prepareStatement(sqlGrupo)
                            stmtGrupo.setInt(1, evento.idEtiqueta)
                            val rsGrupo = stmtGrupo.executeQuery()
                            
                            var idGrupo: Int? = null
                            if (rsGrupo.next()) {
                                val grupoId = rsGrupo.getInt("id_grupo")
                                if (!rsGrupo.wasNull()) {
                                    idGrupo = grupoId
                                }
                            }
                            rsGrupo.close()
                            stmtGrupo.close()
                            
                            // If label belongs to a group, get all group members
                            if (idGrupo != null) {
                                val sqlMembers = "SELECT id_usuario FROM grupos_usuarios WHERE id_grupo = ?"
                                val stmtMembers = conn.prepareStatement(sqlMembers)
                                stmtMembers.setInt(1, idGrupo)
                                val rsMembers = stmtMembers.executeQuery()
                                
                                while (rsMembers.next()) {
                                    grupoUsuarios.add(rsMembers.getInt("id_usuario"))
                                }
                                rsMembers.close()
                                stmtMembers.close()
                            }
                        }
                        
                        // Insert into eventos_usuarios
                        if (grupoUsuarios.isNotEmpty()) {
                            // Event has a group label - add all group members
                            val sqlLink = "INSERT INTO eventos_usuarios (id_evento, id_usuario) VALUES (?, ?)"
                            val stmtLink = conn.prepareStatement(sqlLink)
                            
                            for (userId in grupoUsuarios) {
                                stmtLink.setInt(1, generatedId)
                                stmtLink.setInt(2, userId)
                                stmtLink.addBatch()
                            }
                            
                            stmtLink.executeBatch()
                            stmtLink.close()
                        } else {
                            // No group label - add only the creator
                            val sqlLink = "INSERT INTO eventos_usuarios (id_evento, id_usuario) VALUES (?, ?)"
                            val stmtLink = conn.prepareStatement(sqlLink)
                            stmtLink.setInt(1, generatedId)
                            stmtLink.setInt(2, idUsuario)
                            stmtLink.executeUpdate()
                            stmtLink.close()
                        }
                        
                        conn.commit()
                    }
                    generatedKeys.close()
                } else {
                    conn.rollback()
                }
                
                statement.close()
                conn.autoCommit = true
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        
        generatedId
    }

    override suspend fun insertInvitados(idEvento: Int, emails: List<String>): Boolean = withContext(Dispatchers.IO) {
        val connection = DatabaseConnection.getConnection()
        var success = true
        
        try {
            connection?.let { conn ->
                // Ensure table exists
                val createTableSql = """
                    CREATE TABLE IF NOT EXISTS invitaciones (
                        id_invitacion INT AUTO_INCREMENT PRIMARY KEY,
                        id_evento INT NOT NULL,
                        email VARCHAR(255) NOT NULL,
                        FOREIGN KEY (id_evento) REFERENCES eventos(id_evento) ON DELETE CASCADE
                    )
                """
                val createStmt = conn.createStatement()
                createStmt.execute(createTableSql)
                createStmt.close()
                
                // Insert guests
                val insertSql = "INSERT INTO invitaciones (id_evento, email) VALUES (?, ?)"
                val insertStmt = conn.prepareStatement(insertSql)
                
                for (email in emails) {
                    insertStmt.setInt(1, idEvento)
                    insertStmt.setString(2, email)
                    insertStmt.addBatch()
                }
                
                insertStmt.executeBatch()
                insertStmt.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            success = false
        }
        
        success
    }

    override suspend fun getDiasConEventos(idUsuario: Int, year: Int, month: Int): List<Int> = withContext(Dispatchers.IO) {
        val dias = mutableListOf<Int>()
        val connection = DatabaseConnection.getConnection()
        
        try {
            connection?.let { conn ->
                val sql = """
                    SELECT DISTINCT DAY(e.fecha_inicio) as dia
                    FROM eventos e
                    INNER JOIN eventos_usuarios eu ON e.id_evento = eu.id_evento
                    WHERE eu.id_usuario = ?
                    AND YEAR(e.fecha_inicio) = ?
                    AND MONTH(e.fecha_inicio) = ?
                    ORDER BY dia ASC
                """
                
                val statement = conn.prepareStatement(sql)
                statement.setInt(1, idUsuario)
                statement.setInt(2, year)
                statement.setInt(3, month)
                
                val resultSet = statement.executeQuery()
                
                while (resultSet.next()) {
                    dias.add(resultSet.getInt("dia"))
                }
                
                resultSet.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        
        dias
    }

    override suspend fun getEventosByUsuarioAndRango(idUsuario: Int, fechaInicio: LocalDate, fechaFin: LocalDate): List<Evento> = withContext(Dispatchers.IO) {
        val eventos = mutableListOf<Evento>()
        val connection = DatabaseConnection.getConnection()
        
        try {
            connection?.let { conn ->
                val sql = """
                    SELECT e.*, et.color 
                    FROM eventos e
                    INNER JOIN eventos_usuarios eu ON e.id_evento = eu.id_evento
                    LEFT JOIN etiquetas et ON e.id_etiqueta = et.id_etiqueta
                    WHERE eu.id_usuario = ? 
                    AND e.fecha_inicio >= ? AND e.fecha_inicio <= ?
                    ORDER BY e.fecha_inicio ASC
                """
                
                val statement = conn.prepareStatement(sql)
                statement.setInt(1, idUsuario)
                // Append time to invoke full day range coverage
                statement.setString(2, fechaInicio.atStartOfDay().toString())
                statement.setString(3, fechaFin.atTime(23, 59, 59).toString())
                
                val resultSet = statement.executeQuery()
                
                while (resultSet.next()) {
                    eventos.add(mapResultSetToEvento(resultSet))
                }
                
                resultSet.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        
        eventos
    }

    private fun mapResultSetToEvento(rs: ResultSet): Evento {
        // Handle potential missing columns if join didn't happen (defensive)
        var color: String? = null
        var idEtiqueta: Int? = null
        try {
            color = rs.getString("color")
            idEtiqueta = rs.getObject("id_etiqueta") as? Int
        } catch (e: Exception) {
            // Column might not exist in some queries if not updated? 
            // But we will update all queries.
        }

        return Evento(
            idEvento = rs.getInt("id_evento"),
            idCreador = rs.getObject("id_creador") as? Int,
            idEtiqueta = idEtiqueta,
            titulo = rs.getString("titulo"),
            descripcion = rs.getString("descripcion"),
            fechaInicio = rs.getTimestamp("fecha_inicio").toInstant()
                .atZone(java.time.ZoneId.systemDefault())
                .toLocalDateTime(),
            fechaFin = rs.getTimestamp("fecha_fin").toInstant()
                .atZone(java.time.ZoneId.systemDefault())
                .toLocalDateTime(),
            ubicacion = rs.getString("ubicacion"),
            estado = EstadoEvento.fromString(rs.getString("estado")),
            creadoEl = rs.getTimestamp("creado_el")?.toInstant()
                ?.atZone(java.time.ZoneId.systemDefault())
                ?.toLocalDateTime(),
            actualizadoEl = rs.getTimestamp("actualizado_el")?.toInstant()
                ?.atZone(java.time.ZoneId.systemDefault())
                ?.toLocalDateTime(),
            colorEtiqueta = color
        )
    }

    override suspend fun searchEventos(idUsuario: Int, query: String): List<Evento> = withContext(Dispatchers.IO) {
        val eventos = mutableListOf<Evento>()
        val connection = DatabaseConnection.getConnection()
        
        try {
            connection?.let { conn ->
                val sql = """
                    SELECT e.*, et.color 
                    FROM eventos e
                    INNER JOIN eventos_usuarios eu ON e.id_evento = eu.id_evento
                    LEFT JOIN etiquetas et ON e.id_etiqueta = et.id_etiqueta
                    WHERE eu.id_usuario = ? 
                    AND e.titulo LIKE ?
                    ORDER BY e.fecha_inicio ASC
                """
                
                val statement = conn.prepareStatement(sql)
                statement.setInt(1, idUsuario)
                statement.setString(2, "%$query%")
                
                val resultSet = statement.executeQuery()
                
                while (resultSet.next()) {
                    eventos.add(mapResultSetToEvento(resultSet))
                }
                
                resultSet.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        
        eventos
    }

    override suspend fun getEventoById(idEvento: Int): Evento? = withContext(Dispatchers.IO) {
        var evento: Evento? = null
        val connection = DatabaseConnection.getConnection()
        try {
            connection?.let { conn ->
                val sql = """
                    SELECT e.*, et.color 
                    FROM eventos e
                    LEFT JOIN etiquetas et ON e.id_etiqueta = et.id_etiqueta
                    WHERE e.id_evento = ?
                """
                val statement = conn.prepareStatement(sql)
                statement.setInt(1, idEvento)
                val resultSet = statement.executeQuery()
                if (resultSet.next()) {
                    evento = mapResultSetToEvento(resultSet)
                }
                resultSet.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        evento
    }

    override suspend fun deleteEvento(idEvento: Int): Boolean = withContext(Dispatchers.IO) {
        var success = false
        val connection = DatabaseConnection.getConnection()
        try {
            connection?.let { conn ->
                val sql = "DELETE FROM eventos WHERE id_evento = ?"
                val statement = conn.prepareStatement(sql)
                statement.setInt(1, idEvento)
                val affectedRows = statement.executeUpdate()
                success = affectedRows > 0
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        success
    }
}
